src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -340,6 +352,215 @@ class AutoDiffWhileFwd
+    // pass differentials for all non-constant results or non-constants
+    // operands.
+    SmallVector<bool> operandsActive(orig->getNumOperands(), false);
+    for (int i = 0; i < operandsActive.size(); ++i) {
+      operandsActive[i] = !gutils->isConstantValue(orig->getOperand(i)) ||
+                          !gutils->isConstantValue(orig->getResult(i));
+    }
+
+    // The reverse of the while loop is a for loop where the number
+    // of iterations is either known or cached from the augmented primal.
+    WhileLoopInfo info(cast<WhileOp>(orig));
+    Value numIters =
+        info.computeInfo().succeeded() && info.isConstant()
+            ? makeI64Constant(orig->getLoc(), builder, info.getNumIters())
+            : gutils->popCache(caches[0], builder);
+
+    auto unrankedTensorType = RankedTensorType::get({}, builder.getI64Type());
+    auto iterVar =
+        builder
+            .create<ConstantOp>(
+                orig->getLoc(), unrankedTensorType,
+                SplatElementsAttr::get(unrankedTensorType,
+                                       ArrayRef<Attribute>(IntegerAttr::get(
+                                           builder.getI64Type(), 0))))
+            .getResult();
+
+    SmallVector<Value> operands;
+    operands.reserve(orig->getNumResults() + 1);
+    operands.push_back(iterVar);
+
+    for (auto [active, res] : llvm::zip(operandsActive, orig->getResults())) {
+      if (active) {
+        operands.push_back(gutils->diffe(res, builder));
+        if (!gutils->isConstantValue(res))
+          gutils->zeroDiffe(res, builder);
+      }
+    }
+
+    auto revWhile = builder.create<WhileOp>(
+        orig->getLoc(), ValueRange(operands).getTypes(), operands);
+    auto &condReg = revWhile.getCond();
+    auto &bodyReg = revWhile.getBody();
+
+    auto cond = new Block();
+    auto body = new Block();
+
+    condReg.push_back(cond);
+    bodyReg.push_back(body);
+
+    {
+      for (auto operand : operands) {
+        cond->addArgument(operand.getType(), orig->getLoc());
+      }
+      auto condIterVar = cond->getArgument(0);
+
+      OpBuilder condBuilder(cond, cond->end());
+      condBuilder.create<ReturnOp>(
+          orig->getLoc(),
+          ValueRange(condBuilder
+                         .create<CompareOp>(orig->getLoc(), condIterVar,
+                                            numIters, ComparisonDirection::LT)
+                         .getResult()));
+    }
+
+    bool valid = true;
+    {
+      for (auto operand : operands) {
+        body->addArgument(operand.getType(), orig->getLoc());
+      }
+      OpBuilder bodyBuilder(body, body->end());
+      auto one = bodyBuilder.create<ConstantOp>(
+          orig->getLoc(), unrankedTensorType,
+          SplatElementsAttr::get(unrankedTensorType,
+                                 ArrayRef<Attribute>(IntegerAttr::get(
+                                     bodyBuilder.getI64Type(), 1))));
+      Value bodyIterVar =
```

Comment:
as weird, annoying as it is, we should probably use the createAdd from the autodiff type interface. Even though here we have to raise it back into a stablehlo.add, it does mean that new types will use the right add [if for example not correct in stablehlo.add]. I'm okay with this though if you feel strongly.

Reply:
ah apologies, this add is for a new induction variable.

In this case same comment applies of we should re-use an existing induction variable [almost all cases presently], when possible.

Reply:
Alternatively [or perhaps in addition to], we really should import most of that while op optimization stuff from polygeist [which also does redundant induction variable elimination somewhere iirc]